### PR TITLE
Enables PKB to run on a windows controller

### PR DESCRIPTION
**Pull request by ehankland**

### CODE REVIEW

**Comment by ehankland** [comment]
test this please.

**Comment by ehankland** [comment]
test this please.

**Comment by ehankland** [comment]
test this please.

File: README.md [hunk 1]
```diff
@@ -71,6 +71,34 @@ Before you can run the PerfKit Benchmaker on Cloud providers you need accounts a
 You also need the software dependencies, which are mostly command line tools and credentials to access your
 accounts without a password.  The following steps should help you get the CLI tool auth in place.
 
+If you are running on Windows, you will need to install Github Windows
```
Comment (cmccoy) [inline README.md]: Will a windows controller work for linux guests? If not, we should call that out.

Comment (ehankland) [inline README.md]: Yes a windows controller can run benchmarks on linux guests - that's all this PR enables - windows guests will come later.

Comment (cmccoy) [inline README.md]: Ah - okay, thanks.

File: perfkitbenchmarker/vm_util.py [hunk 2]
```diff
@@ -87,8 +90,15 @@ def PrependTempDir(file_name):
 
 def GenTempDir():
   """Creates the tmp dir for the current run if it does not already exist."""
-  create_cmd = ['mkdir', '-p', GetTempDir()]
-  create_process = subprocess.Popen(create_cmd)
+  if os.name == WINDOWS:
+    create_cmd = ['powershell', 'mkdir', GetTempDir()]
+  else:
+    create_cmd = ['mkdir', '-p', GetTempDir()]
+  shell_value = True if os.name == WINDOWS else False
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Nit: shell_value = os.name == WINDOWS

File: perfkitbenchmarker/vm_util.py [hunk 3]
```diff
@@ -352,7 +368,8 @@ def IssueCommand(cmd, force_info_log=False, suppress_warning=False, env=None):
   full_cmd = ' '.join(cmd)
   logging.info('Running: %s', full_cmd)
 
-  process = subprocess.Popen(cmd, env=env,
+  shell_value = True if os.name == WINDOWS else False
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Introduce a function?

File: perfkitbenchmarker/vm_util.py [hunk 4]
```diff
@@ -475,14 +493,37 @@ def ShouldRunOnInternalIpAddress(sending_vm, receiving_vm):
...
-      ['bash', '-c', 'ls -1t %s | head -1' % TEMP_DIR])
+  if os.name == WINDOWS:
+    cmd = ['powershell', '-Command',
+           'gci %s | sort LastWriteTime | select -last 1' % TEMP_DIR]
+  else:
+    cmd = ['bash', '-c', 'ls -1t %s | head -1' % TEMP_DIR]
+  stdout, _, _ = IssueCommand(cmd)
   try:
-    return regex_util.ExtractGroup('run_(.*)', stdout)
+    return regex_util.ExtractGroup('run_([^\s]*)', stdout)
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Not necessary, to change but `\S` is the complement of `\s`.

File: perfkitbenchmarker/vm_util.py [hunk 5]
```diff
@@ -475,14 +493,37 @@ def ShouldRunOnInternalIpAddress(sending_vm, receiving_vm):
...
   except regex_util.NoMatchError:
     return None
 
 
+@contextlib.contextmanager
+def NamedTempFile(prefix='tmp', suffix='', dir=None):
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Maybe `NamedTemporaryFile`, to match `tempfile`?

File: perfkitbenchmarker/vm_util.py [hunk 6]
```diff
@@ -475,14 +493,37 @@ def ShouldRunOnInternalIpAddress(sending_vm, receiving_vm):
...
+  compatible way. This serves a similar role, but allows the file to be closed
+  within a "with" statement without causing the file to be unlinked until the
+  context exits.
+  """
+  f = tempfile.NamedTemporaryFile(prefix=prefix, suffix=suffix,
+                                  dir=dir, delete=False)
+  try:
+    yield f
+  finally:
+    os.unlink(f.name)
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: What about closing the fd?

File: perfkitbenchmarker/benchmarks/hpcc_benchmark.py [hunk 7]
```diff
@@ -86,13 +86,14 @@ def CreateMachineFile(vms):
...
     vms: The list of vms which will be in the cluster.
   """
-  with tempfile.NamedTemporaryFile() as machine_file:
+  with vm_util.NamedTempFile() as machine_file:
     master_vm = vms[0]
     machine_file.write('localhost slots=%d\n' % master_vm.num_cpus)
     for vm in vms[1:]:
       machine_file.write('%s slots=%d\n' % (vm.internal_ip,
                                             vm.num_cpus))
     machine_file.flush()
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/hpcc_benchmark.py]: May as well drop `.flush()` if you are `close()`ing.

File: perfkitbenchmarker/gcp/gce_virtual_machine.py [hunk 8]
```diff
@@ -83,10 +82,11 @@ def _Create(self):
     super(GceVirtualMachine, self)._Create()
     with open(self.ssh_public_key) as f:
       public_key = f.read().rstrip('\n')
-    with tempfile.NamedTemporaryFile(dir=vm_util.GetTempDir(),
-                                     prefix='key-metadata') as tf:
+    with vm_util.NamedTempFile(dir=vm_util.GetTempDir(),
+                               prefix='key-metadata') as tf:
       tf.write('%s:%s\n' % (self.user_name, public_key))
       tf.flush()
+      tf.close()
```
Comment (cmccoy) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: As above: may as well drop `.flush()` if you are `close()`ing.

File: perfkitbenchmarker/virtual_machine.py [hunk 9]
```diff
@@ -260,9 +261,16 @@ def RemoteCopy(self, file_path, remote_path='', copy_to=True):
     Raises:
       SshConnectionError: If there was a problem copying the file.
     """
+    if os.name == vm_util.WINDOWS:
+      if ':' in file_path:
+        # Scp doesn't like colons in paths.
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Nit: lowercase `scp` to match command name.

File: perfkitbenchmarker/virtual_machine.py [hunk 10]
```diff
@@ -260,9 +261,16 @@ def RemoteCopy(self, file_path, remote_path='', copy_to=True):
     Raises:
       SshConnectionError: If there was a problem copying the file.
     """
+    if os.name == vm_util.WINDOWS:
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Could you introduce a function to determine the OS?

File: perfkitbenchmarker/virtual_machine.py [hunk 11]
```diff
@@ -333,8 +341,11 @@ def RemoteCommand(self, command,
     Raises:
       SshConnectionError: If there was a problem establishing the connection.
     """
+    if os.name == vm_util.WINDOWS:
+      command = command.replace('\n', '\\n')
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Could you add a comment on why this is needed?

File: perfkitbenchmarker/vm_util.py [hunk 12]
```diff
@@ -87,8 +90,15 @@ def PrependTempDir(file_name):
 
 def GenTempDir():
   """Creates the tmp dir for the current run if it does not already exist."""
-  create_cmd = ['mkdir', '-p', GetTempDir()]
-  create_process = subprocess.Popen(create_cmd)
+  if os.name == WINDOWS:
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Consider using python functions here. `os.path.exists` and [`os.makedirs`](https://docs.python.org/2/library/os.html#os.makedirs) work across platforms.

File: README.md (hunk 1 above)
Comment (cmccoy) [inline README.md]: Nit: Git**H**ub

File: README.md [hunk 13]
```diff
@@ -71,6 +71,34 @@ Before you can run the PerfKit Benchmaker on Cloud providers you need accounts a
...
+since it includes tools like openssl and an ssh client. Alternatively you can
+install Cygwin since it should include the same tools.
+
+## Install python 2 (and pip)
```
Comment (cmccoy) [inline README.md]: Nit: captialize Python. Recommend specifying Python 2.7 as well - we use features not available in earlier versions.

File: README.md [hunk 14]
```diff
@@ -71,6 +71,34 @@ Before you can run the PerfKit Benchmaker on Cloud providers you need accounts a
...
+using your distribution's package manager.
+
+If you just need to install pip, go [here](http://pip.readthedocs.org/en/latest/installing.html).
+On Windows follow the instructions on the page above.
+On Debian/Ubuntu you can run the command below (or the equivalent command on different distributions):
+```
+$ sudo apt-get install python-pip -y
+```
+
+## (*Windows Only*) Install Github Windows
```
Comment (cmccoy) [inline README.md]: Nit: capital H in GitHub.

File: README.md [hunk 15]
```diff
@@ -71,6 +71,34 @@ Before you can run the PerfKit Benchmaker on Cloud providers you need accounts a
...
+
+Instructions: https://windows.github.com/
+
+Make sure that openssl/ssh/scp/ssh-keygen are on your path (you will need to update the PATH environment variable).
+The path to these commands should be
+
+"C:\\Users\\\<user\>\\AppData\\Local\\GitHub\\PortableGit\_\<guid\>\\bin".
```
Comment (cmccoy) [inline README.md]: Nit: backquote this so that it renders fixed width.

**Comment by cmccoy** [comment]
Looks good. Some small comments.

Given that Windows requires some additional prerequisites to be installed, I wonder if we should do a pre-run check for the presence of:

* `ssh`
* `ssh-keygen`
* `scp`
* `openssl`

File: README.md (hunk 1 above)
Comment (ehankland) [inline README.md]: Done

File: perfkitbenchmarker/vm_util.py (hunk 2 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

File: perfkitbenchmarker/vm_util.py (hunk 3 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

File: perfkitbenchmarker/vm_util.py (hunk 4 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Ack.

File: perfkitbenchmarker/vm_util.py (hunk 5 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

File: perfkitbenchmarker/vm_util.py (hunk 6 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

File: perfkitbenchmarker/benchmarks/hpcc_benchmark.py (hunk 7 above)
Comment (ehankland) [inline perfkitbenchmarker/benchmarks/hpcc_benchmark.py]: Done

File: perfkitbenchmarker/gcp/gce_virtual_machine.py (hunk 8 above)
Comment (ehankland) [inline perfkitbenchmarker/gcp/gce_virtual_machine.py]: Done

File: perfkitbenchmarker/virtual_machine.py (hunk 9 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done

File: perfkitbenchmarker/virtual_machine.py (hunk 10 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done

File: perfkitbenchmarker/virtual_machine.py (hunk 11 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done

File: perfkitbenchmarker/vm_util.py (hunk 12 above)
Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

File: README.md (hunk 13 above)
Comment (ehankland) [inline README.md]: Done

File: README.md (hunk 14 above)
Comment (ehankland) [inline README.md]: Done

File: README.md (hunk 15 above)
Comment (ehankland) [inline README.md]: Done

**Comment by ehankland** [comment]
PTAL

File: perfkitbenchmarker/vm_util.py [hunk 16]
```diff
@@ -512,6 +546,28 @@ def GenerateSSHConfig(vms):
...
+  """Return True if the given executable can be found on the path."""
+  cmd = ['where'] if RunningOnWindows() else ['which']
+  cmd.append(executable_name)
+
+  shell_value = RunningOnWindows()
+  process = subprocess.Popen(cmd,
+                             shell=shell_value,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE)
+  process.wait()
```
Comment (cmccoy) [inline perfkitbenchmarker/vm_util.py]: Unlikely given the commands used, but this [risks deadlock](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait) if `cmd` produces a lot of output. Prefer check_output, communicate, or opening a handle to `os.devnull` and using it for stdout and stderr.

Comment (ehankland) [inline perfkitbenchmarker/vm_util.py]: Done

**Pull request merged**